src/gainmap.c

```diff
@@ -286,4 +299,327 @@ avifResult avifImageApplyGainMap(const avifImage * baseImage,
+        gainMapMetadata.gainMapMin[c] = gainMapMinLog2[singleChannel ? 0 : c];
+        gainMapMetadata.gainMapMax[c] = gainMapMaxLog2[singleChannel ? 0 : c];
+        gainMapMetadata.baseHdrHeadroom = avifRoundf(log2f(baseMax));
+        gainMapMetadata.alternateHdrHeadroom = avifRoundf(log2f(altMax));
+        // baseOffset, alternateOffset and gainMapGamma are all left to their default values.
+        // They could be tweaked based on the images to optimize quality/compression.
+    }
+    if (!avifGainMapMetadataDoubleToFractions(&gainMap->metadata, &gainMapMetadata)) {
+        res = AVIF_RESULT_UNKNOWN_ERROR;
+        goto cleanup;
+    }
+
+    // Scale the gain map values to map [min, max] range to [0, 1].
+    for (int c = 0; c < 3; ++c) {
+        const float range = gainMapMaxLog2[c] - gainMapMinLog2[c];
+        if (range <= 0.0f) {
+            continue;
+        }
+
+        for (int j = 0; j < height; ++j) {
+            for (int i = 0; i < width; ++i) {
+                // Remap [min; max] range to [0; 1]
+                const float v = AVIF_CLAMP(gainMapF[c][j * width + i], gainMapMinLog2[c], gainMapMaxLog2[c]);
+                gainMapF[c][j * width + i] = powf((v - gainMapMinLog2[c]) / range, (float)gainMapMetadata.gainMapGamma[c]);
+            }
+        }
+        if (singleChannel) {
+            break;
+        }
+    }
+
+    // Convert the gain map to YUV.
+    const uint32_t requestedWidth = gainMapImage->width;
+    const uint32_t requestedHeight = gainMapImage->height;
+    gainMapImage->width = width;
+    gainMapImage->height = height;
+
+    avifImageFreePlanes(gainMapImage, AVIF_PLANES_ALL); // Free planes in case they were already allocated.
+    res = avifImageAllocatePlanes(gainMapImage, AVIF_PLANES_YUV);
+    if (res != AVIF_RESULT_OK) {
+        goto cleanup;
+    }
+
+    avifRGBImageSetDefaults(&gainMapRGB, gainMapImage);
+    res = avifRGBImageAllocatePixels(&gainMapRGB);
+    if (res != AVIF_RESULT_OK) {
+        goto cleanup;
+    }
+
+    avifRGBColorSpaceInfo gainMapRGBInfo;
+    if (!avifGetRGBColorSpaceInfo(&gainMapRGB, &gainMapRGBInfo)) {
+        avifDiagnosticsPrintf(diag, "Unsupported RGB color space");
+        return AVIF_RESULT_NOT_IMPLEMENTED;
+    }
+    for (int j = 0; j < height; ++j) {
+        for (int i = 0; i < width; ++i) {
+            const int offset = j * width + i;
+            const float r = gainMapF[0][offset];
+            const float g = singleChannel ? r : gainMapF[1][offset];
+            const float b = singleChannel ? r : gainMapF[2][offset];
+            const float rgbaPixel[4] = { r, g, b, 1.0f };
+            avifSetRGBAPixel(&gainMapRGB, i, j, &gainMapRGBInfo, rgbaPixel);
+        }
+    }
+
+    res = avifImageRGBToYUV(gainMapImage, &gainMapRGB);
+    if (res != AVIF_RESULT_OK) {
+        goto cleanup;
+    }
+
+    // Scale down the gain map if requested.
+    // Another way would be to scale down the source images, but it seems to perform worse.
```

Comment:
This is "probably down" rather than "always down" here, right?

Reply:
Indeed.